x/local-proxy/main.go

```diff
@@ -0,0 +1,136 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/base64"
+	"flag"
+	"fmt"
+	"log"
+	"net/url"
+	"strings"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/Jigsaw-Code/outline-sdk/transport/shadowsocks"
+	"github.com/Jigsaw-Code/outline-sdk/transport/socks5"
+	"github.com/Jigsaw-Code/outline-sdk/x/local-proxy/proxy/httpproxy"
+	"os"
+	"os/signal"
+	"time"
+)
+
+func makeStreamDialer(transportConfig string) (transport.StreamDialer, error) {
+	if transportConfig == "" {
+		return &transport.TCPStreamDialer{}, nil
+	}
+
+	accessKeyURL, err := url.Parse(transportConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse access key: %w", err)
+	}
+	switch accessKeyURL.Scheme {
+
+	case "socks5":
+		return socks5.NewStreamDialer(&transport.TCPEndpoint{Address: accessKeyURL.Host})
+
+	case "ss":
+		cipherInfoBytes, err := base64.URLEncoding.WithPadding(base64.NoPadding).DecodeString(accessKeyURL.User.String())
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode cipher info [%v]: %v", accessKeyURL.User.String(), err)
+		}
+		cipherName, secret, found := strings.Cut(string(cipherInfoBytes), ":")
+		if !found {
+			return nil, fmt.Errorf("invalid cipher info: no ':' separator")
+		}
+		cryptoKey, err := shadowsocks.NewEncryptionKey(cipherName, secret)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create cipher: %w", err)
+		}
+
+		dialer, err := shadowsocks.NewStreamDialer(&transport.TCPEndpoint{Address: accessKeyURL.Host}, cryptoKey)
+		if err != nil {
+			return nil, err
+		}
+
+		prefixStr := accessKeyURL.Query().Get("prefix")
+		if len(prefixStr) > 0 {
+			prefix, err := parseStringPrefix(prefixStr)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse prefix: %w", err)
+			}
+			dialer.SaltGenerator = shadowsocks.NewPrefixSaltGenerator(prefix)
+		}
+		return dialer, nil
+
+	default:
+		return nil, fmt.Errorf("access key scheme %v:// is not supported", accessKeyURL.Scheme)
+	}
+}
+
+func parseStringPrefix(utf8Str string) ([]byte, error) {
+	runes := []rune(utf8Str)
+	rawBytes := make([]byte, len(runes))
+	for i, r := range runes {
+		if (r & 0xFF) != r {
+			return nil, fmt.Errorf("character out of range: %d", r)
+		}
+		rawBytes[i] = byte(r)
+	}
+	return rawBytes, nil
+}
+
+func main() {
+	transportFlag := flag.String("transport", "", "Transport config")
+	addrFlag := flag.String("addr", "localhost:54321", "Local proxy addr")
+	flag.Parse()
+
+	dialer, err := makeStreamDialer(*transportFlag)
+	if err != nil {
+		log.Fatalf("Could not create dialer: %+v\n", err)
+	}
+
+	proxy := httpproxy.NewConnectHandler(dialer)
+	proxy.OnError = httpproxy.OnError
+
+	if err := proxy.StartServer(*addrFlag); err != nil {
+		log.Fatal("Proxy server failed")
+	}
+
+	println("Proxy server started on ", proxy.GetAddr())
+
+	// wait for kill/interrupt signal to stop the proxy with a timeout of 5 seconds.
```

Comment:
We prefer comments as sentences with proper capitalization.
```suggestion
	// Wait for kill/interrupt signal to stop the proxy with a timeout of 5 seconds.
```